### PR TITLE
enhancement(cli): Consolidate and beautify `validate` 

### LINKED ISSUE #2377: Validate topology incorrectly reports topology errors when using multiple config files
Running `vector validate --topology` on more than one config fails if the topology is spread across multiple files. Interestingly, the errors are different based on the order of the config files supplied. Vector happily runs these config files, so `vector validate --topology` incorrectly reports that the config files aren't compatible.

### Repro

```sh
vector --version
vector 0.9.0-nightly (gdfe4620 x86_64-unknown-linux-musl 2020-04-14)
```

```toml
# a.toml
[sources.a_journald_src]
  type = "journald"
  units = []

[sinks.a_blackhole]
  type = "blackhole"
  inputs = ["b_lua"]
  print_amount = 1000
```

```toml
# b.toml
[transforms.b_lua]
  type = "lua"
  version = "2"
  inputs = ["a_journald_src"]
  hooks.process = "function (event, emit) emit(event) end"
```

```sh
vector validate --topology --deny-warnings a.toml b.toml

ERROR vector: Topology error: Input "b_lua" for sink "a_blackhole" doesn't exist.
ERROR vector: Failed to verify config file topology. path="a.toml
```

```sh
vector validate --topology --deny-warnings b.toml a.toml

ERROR vector: Topology error: Input "journald_src" for transform "all_sources_for_service" doesn't exist.
ERROR vector: Failed to verify config file topology. path="b.toml"
```

### PR DESCRIPTION
**Pull request by ktff**
Closes #2377, #2278

Notes for `validate`: 
* Accepts `--no_healthcheck`, `--no_environment`.
* Has shorthands `-nte`, `-net`, `-nt`, and `-ne`.
* Everything valid has shorter print form.
* Everything invalid has longer print form.
* Continues validation in spite of errors wherever it's possible.
* Plays nicely with `verbose`, `quiet`, and `color` flags.

### Current look 
<details>
<summary>Examples</summary>

---

```
     Running `target/debug/vector validate -nte ./../validate/1.toml`
√ Loaded "../validate/1.toml"
-----------------------------
                    Validated
```
---

```
     Running `target/debug/vector validate -ne ./../validate/1.toml`
√ Loaded "../validate/1.toml"
√ Configuration topology
-----------------------------
                    Validated
```
---

```
     Running `target/debug/vector validate ./../validate/1.toml`
√ Loaded "../validate/1.toml"
√ Configuration topology
√ Component configuration
May 30 14:10:23.173 ERROR vector::topology::builder: Healthcheck: Failed Reason: Connect error: Connection refused (os error 111)
x Health check for `out` failed

```
---

```
     Running `target/debug/vector validate  ./../validate/2.toml`
Failed to parse "../validate/2.toml"
------------------------------------
x unknown variant `jabberwocky`, expected one of `aws_cloudwatch_logs`, `aws_cloudwatch_metrics`, `aws_kinesis_firehose`, `aws_kinesis_streams`, `aws_s3`, `blackhole`, `clickhouse`, `console`, `datadog_logs`, `datadog_metrics`, `elasticsearch`, `file`, `gcp_cloud_storage`, `gcp_pubsub`, `gcp_stackdriver_logs`, `honeycomb`, `http`, `humio_logs`, `influxdb_metrics`, `kafka`, `logdna`, `loki`, `new_relic_logs`, `papertrail`, `prometheus`, `pulsar`, `sematext`, `socket`, `splunk_hec`, `statsd`, `vector` for key `sinks.out`

```
---

```
     Running `target/debug/vector validate -nte ./../validate/3.toml`
√ Loaded "../validate/3.toml"
-----------------------------
                    Validated

```
---

```
     Running `target/debug/vector validate -ne ./../validate/3.toml`
√ Loaded "../validate/3.toml"
√ Configuration topology
-----------------------------
                    Validated

```
---

```
     Running `target/debug/vector validate ./../validate/3.toml`
√ Loaded "../validate/3.toml"
√ Configuration topology
√ Component configuration
√ Health check `out`
-----------------------------
                    Validated

```
---

```
     Running `target/debug/vector validate ./../validate/4.toml`
√ Loaded "../validate/4.toml"
√ Configuration topology
√ Component configuration
May 30 14:10:25.702 ERROR vector::topology::builder: Healthcheck: Failed Reason: Connect error: Connection refused (os error 111)
x Health check for `out2` failed
May 30 14:10:25.703 ERROR vector::topology::builder: Healthcheck: Failed Reason: Connect error: Connection refused (os error 111)
x Health check for `out1` failed

```
---

```
     Running `target/debug/vector validate ./../validate/5.toml`
√ Loaded "../validate/5.toml"
√ Configuration topology

Component errors
----------------
x Sink "out1": Must set either 'region' or 'endpoint'
x Sink "out2": Not a valid AWS region: moonbase-alpha
x Sink "out3": Only one of 'region' or 'endpoint' can be specified
x Sink "out4": Failed to parse custom endpoint as URI: invalid uri character

```
---

```
     Running `target/debug/vector validate ./../validate/6.toml`
√ Loaded "../validate/6.toml"

Topology errors
---------------
x Cyclic dependency detected in the chain [ four -> two -> three -> four ]

√ Component configuration
May 30 14:10:27.408 ERROR vector::topology::builder: Healthcheck: Failed Reason: Connect error: Connection refused (os error 111)
x Health check for `out` failed

```
---

```
     Running `target/debug/vector validate ./../validate/7.toml`
√ Loaded "../validate/7.toml"

Topology errors
---------------
x Cyclic dependency detected in the chain [ four -> two -> three -> four ]

√ Component configuration
~ Health check disabled for `out`


   ```
---

```
     Running `target/debug/vector validate ./../validate/8.toml`
Failed to parse "../validate/8.toml"
------------------------------------
x unexpected character found: `\\` at line 1

```
---

```
     Running `target/debug/vector validate ./../validate/9.toml`
Failed to parse "../validate/9.toml"
------------------------------------
x missing field `mode` for key `sources.in`

```
---

```
     Running `target/debug/vector validate -nte ./../validate/10.toml`
√ Loaded "../validate/10.toml"
------------------------------
                     Validated

```
---

```
     Running `target/debug/vector validate -ne ./../validate/10.toml`
√ Loaded "../validate/10.toml"

Topology errors
---------------
x Input "asdf" for sink "out" doesn't exist.
x Input "qwerty" for transform "sampler" doesn't exist.


```
---

```
     Running `target/debug/vector validate ./../validate/10.toml`
√ Loaded "../validate/10.toml"

Topology errors
---------------
x Input "asdf" for sink "out" doesn't exist.
x Input "qwerty" for transform "sampler" doesn't exist.

√ Component configuration
May 30 14:10:30.803 ERROR vector::topology::builder: Healthcheck: Failed Reason: Connect error: Connection refused (os error 111)
x Health check for `out` failed

```
---

```
     Running `target/debug/vector validate ./../validate/11.toml`
√ Loaded "../validate/11.toml"

Topology warnings
-----------------
~ Transform "sampler2" has no consumers
~ Source "in2" has no consumers

√ Configuration topology
√ Component configuration
May 30 14:10:31.652 ERROR vector::topology::builder: Healthcheck: Failed Reason: Connect error: Connection refused (os error 111)
x Health check for `out` failed

```
---

```
     Running `target/debug/vector validate ./../validate/3.toml ./../validate/4.toml`
√ Loaded "../validate/3.toml"

Failed to merge config "../validate/4.toml"
-------------------------------------------
x duplicate source name found: in

√ Configuration topology
√ Component configuration
√ Health check `out`

```
</details>

---
With difference that console has color green on `√`, red on `x`, yellow on `~`, and green on `Validated`. 

### Open questions/tasks

- [x] We could also pretty print errors for health checks but that would require extending `Task`. (EDIT: No, too intrusive)
- [x] We could also pretty print errors for missing config files. (EDIT: it could be considered that this is a part of validating command options so it's fine for it to remain hard error)
- [x] Explore if it would be possible to move formatting to logging infrastructure. This would simplify the code and would also pretty print other logs which includes health checks errors. (EDIT: No, too intrusive)
- [x] Should we have some shorthand for minimal check? (ref. https://github.com/timberio/vector/pull/2622#pullrequestreview-415190754)
- [x] Update documentation
     - [x] Remove `dry_run`
     - [x] Update `validate` command

### CODE REVIEW

**Comment by ktff** [comment]
> With --no-topology I'm still getting errors like:
>
>$ vector validate --no-topology --no-healthchecks ./tests/behavior/transforms/filter.toml 
>- Component errors
>    x No sources defined in the config.

Fixed. And #2377 is also fixed.

---

@binarylogic @Jeffail  What do you think about the style?

**Comment by binarylogic** [comment]
Thanks, @Jeffail mind taking another look?

**Comment by Jeffail** [comment]
The style looks very nice, one nit I'd maybe change is collapsing the healthchecks so:

```
vector validate ./config/examples/stdio.toml 
√ Loaded "config/examples/stdio.toml"
√ Configuration options

Health checks
-------------
√ `out`

√ Configuration topology
-------------------------------------
                            Validated
```

Becomes

```
vector validate ./config/examples/stdio.toml 
√ Loaded "config/examples/stdio.toml"
√ Configuration options
√ Healthcheck `out`
√ Configuration topology
-------------------------------------
                            Validated
```

Ideally if everything is green then I want the output to be as clear and condensed as possible.

Also I'm noticing it's always checking for `data_dir`:

```
vector validate --no-topology --no-healthchecks ./config/examples/file_to_prometheus.toml 
√ Loaded "config/examples/file_to_prometheus.toml"

Component errors
----------------
x Source "file": data_dir "/var/lib/vector" does not exist
```

I'd want at least some way of skipping this check, perhaps with `--no-healthchecks` but I can live with another flag as it's not clearly related.

**Comment by ktff** [comment]
> collapsing the healthchecks

@Jeffail  Then is it acceptable for failing health checks to also be inline, like so:
```
√ Loaded "../topology_panic_conifg.toml"
√ Loaded "../validate_test.toml"
√ Configuration options
√ Health check `s3_archives_haproxy`
√ Health check `datadog_logs_haproxy`
May 25 00:42:03.326 ERROR vector::topology::builder: Healthcheck: Failed Reason: Connect error: Connection refused (os error 111)
x Health check for `out2` failed
√ Health check `s3_archives_nginx`
May 25 00:42:03.327 ERROR vector::topology::builder: Healthcheck: Failed Reason: Connect error: Connection refused (os error 111)
x Health check for `out1` failed
√ Health check `datadog_logs_nginx`
```

They are somewhat inseparable because of the timing of the logs like the above `ERROR`. Because of them we are doing health checks in series so it's easier for the user to determine which logs belong to which health check. 

Alternatively we could try to capture all of the logs and assign them to each check, but it's kinda intrusive and complex so we should try to avoid it. 

---
>Also I'm noticing it's always checking for data_dir:
>
> ```
>vector validate --no-topology --no-healthchecks ./config/examples/file_to_prometheus.toml 
>√ Loaded "config/examples/file_to_prometheus.toml"
>
>Component errors
>----------------
>x Source "file": data_dir "/var/lib/vector" does not exist
>```
>
>I'd want at least some way of skipping this check, perhaps with --no-healthchecks but I can live with another flag as it's not clearly related.

Those are the errors coming from `build` methods of the components which we need to go trough for topology checks and health checks. As a consequence, without those checks, only the config checks can be performed. So I see two ways of addressing this:
 - Add the flag which would disable all checks except for the config checks.
    - Something like `--only-config` or `--no-components`.
 - Don't do those checks if both `--no-healthchecks` and `--no-topology` are used.

For `--config-only` validating step by step isn't so smooth:
 ```
vector validate --only-config
vector validate --no-healthchecks --no-topology
vector validate --no-topology
vector validate 
```
as other option
 ```
vector validate --no-healthchecks --no-topology
vector validate --no-topology
vector validate 
```
but it does give more control.

---
Related to shorthands, what about:
`-n` followed with `t` for `--no-topology`,`h`  for --no-healthchecks, `o` for `--only-config`, in arbitrary order. example. `-nht`, `-nt`, `-no`, `-noht`, etc.

**Comment by Jeffail** [comment]
> @Jeffail Then is it acceptable for failing health checks to also be inline, like so:
> 
> ```
> √ Loaded "../topology_panic_conifg.toml"
> √ Loaded "../validate_test.toml"
> √ Configuration options
> √ Health check `s3_archives_haproxy`
> √ Health check `datadog_logs_haproxy`
> May 25 00:42:03.326 ERROR vector::topology::builder: Healthcheck: Failed Reason: Connect error: Connection refused (os error 111)
> x Health check for `out2` failed
> √ Health check `s3_archives_nginx`
> May 25 00:42:03.327 ERROR vector::topology::builder: Healthcheck: Failed Reason: Connect error: Connection refused (os error 111)
> x Health check for `out1` failed
> √ Health check `datadog_logs_nginx`
> ```
> 
> They are somewhat inseparable because of the timing of the logs like the above `ERROR`. Because of them we are doing health checks in series so it's easier for the user to determine which logs belong to which health check.

Ah okay, in that case I'd personally prefer the inline approach for both.

> Alternatively we could try to capture all of the logs and assign them to each check, but it's kinda intrusive and complex so we should try to avoid it.

Yeah agreed, this is already a huge improvement so no need to dive in deep.

> Those are the errors coming from `build` methods of the components which we need to go trough for topology checks and health checks. As a consequence, without those checks, only the config checks can be performed.

I think it's worth keeping the options here as simple as possible. Given the current flags on offer as a user I would expect the following:

- `--no-topology --no-healthchecks` -> Check that my config doesn't have typos or broken TOML, that config fields are "correct", but doesn't check _anything_ in the context of my environment (including `data_dir`, as I might be running my checks in an environment that I don't have write access to).
- `--no-healthchecks` -> Check the above and also that the topology of my components is correct.
- No flags -> Check all of the above and that all components in my config are able to hit their targets, including any components that interact with the filesystem.

In reality I know that a lot of fields are only validated during `build` and so it's not possible to follow this exactly. However, I'd say making `--no-topology --no-healthchecks` perform a reduced set of checks would be preferred over adding a third flag.

> Related to shorthands, what about:
> `-n` followed with `t` for `--no-topology`,`h` for --no-healthchecks, `o` for `--only-config`, in arbitrary order. example. `-nht`, `-nt`, `-no`, `-noht`, etc.

:+1: That seems good to me.

**Comment by ktff** [comment]
>  * --no-topology --no-healthchecks -> Check that my config doesn't have typos or broken TOML, that config fields are "correct", but doesn't check anything in the context of my environment (including data_dir, as I might be running my checks in an environment that I don't have write access to).
>
> *  --no-healthchecks -> Check the above and also that the topology of my components is correct.
>
> *  No flags -> Check all of the above and that all components in my config are able to hit their targets, including any components that interact with the filesystem.

Then in the line with the above, there is another possibility. We can perform component checks only if we are doing health checks. The topology check can actually be done without building the components, the final topology subcheck that was using the components actually can't fail so we can skip it.

In the end it would look something like:
```
                  no_healthcheck
                             v
                             |------- component_check <------- health_check
config_check <---------------|
                             |------- topology_check
                             ^
                 no_topologycheck
```

But with that, `no_componentcheck` seems like better naming than `no_healthcheck`, or maybe `no_environment`,or `no_environmentcheck`?

@Jeffail what do you think?

**Comment by Jeffail** [comment]
:+1:

> But with that, `no_componentcheck` seems like better naming than `no_healthcheck`, or maybe `no_environment`,or `no_environmentcheck`?
> 
> @Jeffail what do you think?

I'd say `no_environment`.

**Comment by binarylogic** [comment]
This is looking great. Nice work.

**Comment by ktff** [comment]
@Jeffail @lukesteensen this is ready for a review.

File: src/validate.rs [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+    no_topology: bool,
+
+    /// Disables environment checks. That includes component checks and health checks.
+    #[structopt(long)]
+    no_environment: bool,
+
+    /// Shorthand for `--no_topology` and `--no_environment` flags. Just `-n` won't disable anything,
+    /// it needs to be used with `t` for `--no_topology`, and or `e` for `--no_environment` in any order.
+    /// Example:
+    /// `-nte` and `-net` both mean `--no_topology` and `--no_environment`
```
Comment (Jeffail) [inline src/validate.rs:26]: Nit: the actual flags are `--no-topology` and `--no-environment`.

Comment (ktff) [reply src/validate.rs:26]: Good catch.

**Comment by binarylogic** [comment]
@ktff feel free to merge when ready.

**Comment by ktff** [comment]
@binarylogic what about reviewing website changes? They were added in the last three commits.

**Pull request merged**